framework/architecture/hibernate/listeners.rst

```diff
@@ -0,0 +1,129 @@
+This is done in ``AbstractHibernateEntity#internalSetValue()`` which is called by both ``Entity#setValue()`` and the
+generated setter method.
+
+``entityRelationChanging`` event
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This event is fired every time a relation is modified (i.e. if an entity is added to or removed from a relation).
+This is done in ``AbstractRelationAdapter#fireRelationChangingEvent()`` which in turn is called by the
+:java:ref:`ToOneRelationAdapter<ch.tocco.nice2.persist.hibernate.pojo.relation.ToOneRelationAdapter>` and the
+:java:ref:`ToManyRelationAdapter<ch.tocco.nice2.persist.hibernate.pojo.relation.ToManyRelationAdapter>`.
+An event is also always fired for the reverse side of the relation.
+
+See the chapter :doc:`abstract-pojo-entity` for a description when the ``adjusting`` flag is set to true.
+
+``entityReceivedValues`` event
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This event is fired when a newly created entity receives its primary key after it was inserted into the database.
+See ``EntityTransactionContextImpl#executeEntityOperations``.
+This event is rarely used and mostly implemented for compatibility reasons. New listeners should not use this event
+anymore.
+
+EntityListener
+--------------
+
+An :java:ref:`EntityListener<ch.tocco.nice2.persist.entity.events.EntityListener>` is used to listen
+to 'after-commit' events. This means that these listeners are only called after a transaction has been
+successfully committed.
+
+For this purpose we use hibernate's :java:extdoc:`PostCommitInsertEventListener<org.hibernate.event.spi.PostCommitInsertEventListener>`,
+:java:extdoc:`PostCommitUpdateEventListener<org.hibernate.event.spi.PostCommitUpdateEventListener>` and
+:java:extdoc:`PostCommitDeleteEventListener<org.hibernate.event.spi.PostCommitDeleteEventListener>`.
+
+These listener interfaces are implemented by :java:ref:`AfterCommitListenerImpl<ch.tocco.nice2.persist.hibernate.listener.AfterCommitListenerImpl>`,
+which delegates the hibernate events to the corresponding :java:ref:`EntityListener<ch.tocco.nice2.persist.entity.events.EntityListener>`.
+This class is bound to the hibernate events ``POST_COMMIT_INSERT``, ``POST_COMMIT_UPDATE`` and ``POST_COMMIT_DELETE``
+(see :java:ref:`HibernateCoreBootstrapContribution<ch.tocco.nice2.persist.hibernate.bootstrap.HibernateCoreBootstrapContribution>`).
+
+Listeners can either be contributed as hivemind services or registered temporarily through the ``Context`` or ``EntityManager``
+(same as the entity facade listener).
+
+Hibernate does not fire a ``POST_COMMIT_UPDATE`` for an entity if the only change is in a collection and this collection is not the owning side of the association.
+For this special use case there is the :java:ref:`CustomFlushEntityEventListener<ch.tocco.nice2.persist.hibernate.listener.CustomFlushEntityEventListener>`.
+This is class is bound to the hibernate events ``FLUSH_ENTITY`` and checks every entity in the persistence context whether we
```

Comment:
whether ~~we need to fire this event~~ this event needs to be fired manually.

Just for consistency, no first person used elsewhere.